core/src/code.rs

```diff
@@ -290,78 +311,140 @@ fn check_and_canonize_gear_stack_end(module: &mut Module) -> Result<(), CodeErro
+    Export,
+    /// Start section.
+    #[display(fmt = "Start section")]
+    Start,
+}
+
+/// Section error in WASM module.
+#[derive(Debug, PartialEq, Eq, derive_more::Display)]
+pub enum SectionError {
+    /// Section not found.
+    #[display(fmt = "{_0} not found")]
+    NotFound(SectionName),
+    /// Section not supported.
+    #[display(fmt = "{_0} not supported")]
+    NotSupported(SectionName),
+}
+
+/// Memory error in WASM module.
+#[derive(Debug, PartialEq, Eq, derive_more::Display)]
+pub enum MemoryError {
+    /// Memory entry not found in import section.
     #[display(fmt = "Memory entry not found")]
-    MemoryEntryNotFound,
-    /// The provided code doesn't contain export section.
-    #[display(fmt = "Export section not found")]
-    ExportSectionNotFound,
+    EntryNotFound,
+    /// The WASM module has invalid count of static memory pages.
+    #[display(fmt = "The WASM module has invalid count of static memory pages")]
+    InvalidStaticPageCount,
+}
+
+/// Stack end error in WASM module.
+#[derive(Debug, PartialEq, Eq, derive_more::Display)]
+pub enum StackEndError {
+    /// Can't insert new global due to index overflow in global section.
+    #[display(fmt = "Can't insert new global due to index overflow")]
+    GlobalIndexOverflow,
+    /// Pointer to the stack end overlaps data segment.
+    #[display(fmt = "Pointer to the stack end overlaps data segment")]
+    StackEndOverlaps,
+}
+
+/// Initialization error in WASM module.
+#[derive(Debug, PartialEq, Eq, derive_more::Display)]
+pub enum InitializationError {
+    /// Unsupported initialization of gear stack end global variable.
+    #[display(fmt = "Unsupported initialization of gear stack end global variable")]
+    StackEnd,
+    /// Unsupported initialization of data segment.
+    #[display(fmt = "Unsupported initialization of data segment")]
+    DataSegment,
+}
+
+/// Export error in WASM module.
+#[derive(Debug, PartialEq, Eq, derive_more::Display)]
+pub enum ExportError {
+    /// Incorrect global export index. Can occur when export refers to not existing global index.
+    #[display(fmt = "Global index `{_0}` in export index `{_1}` is incorrect")]
+    IncorrectGlobalIndex(u32, u32),
+    /// Exporting mutable globals is restricted by the Gear protocol.
+    #[display(fmt = "Global index `{_0}` in export index `{_1}` cannot be mutable")]
+    MutableGlobalExport(u32, u32),
+    /// Export references to an import function, which is not allowed.
+    #[display(fmt = "Export index `{_0}` references to imported function with index `{_1}`")]
+    ExportReferencesToImport(u32, u32),
+    /// The signature of an exported function is invalid.
+    #[display(fmt = "Exported function with index `{_0}` must have signature `fn f() {{ ... }}`")]
+    InvalidExportFnSignature(u32),
+    /// The provided code contains unnecessary function export.
+    #[display(fmt = "Unnecessary export with index `{_0}` found")]
+    UnnecessaryExport(u32),
```

Comment:
most likely forbidden / unknown ?

Reply:
I think the word "unnecessary" exactly describes error because the "necessary" export is `init` or `handle`

Reply:
Excess mb?

Reply:
It feels weird to return an error for something that was not necessary